### ISSUE TITLE
Unable to sample (NUTS) from `tfd.Mixture` in a graph mode with vectorized `log_prob`

### ISSUE THREAD

**Issue by rrkarim** [tensorflow/probability#684]
I'm trying to sample from `tfd.Mixture` using NUTS sampler and vectorized `log_prob`, everything in graph mode. Here is the snippet, define model:
```python
import tensorflow as tf
import tensorflow_probability as tfp
tfd = tfp.distributions
mix = 0.3
bimix_gauss = tfd.Mixture(
  cat=tfd.Categorical(probs=[mix, 1.-mix]),
  components=[
    tfd.Normal(loc=-1., scale=0.1),
    tfd.Normal(loc=+1., scale=0.5),
    ],
    use_static_graph=True
)
```
`hmc` works just fine:
```python
likelihood = bimix_gauss
def vectorize(logpfn):
    def vectorize_logpfn(*state):
        return tf.vectorized_map(lambda mini_state: logpfn(*mini_state), state)
    return vectorize_logpfn

logpfn = likelihood.log_prob
logpfn = vectorize(logpfn)

state = tf.constant([0.])
@tf.function(autograph=False)
def f(state_):
    states = tfp.mcmc.sample_chain(
        num_results=2,
        num_burnin_steps=2,
        current_state=state_,
        kernel=tfp.mcmc.HamiltonianMonteCarlo(
          target_log_prob_fn=logpfn,
          step_size=0.5,
          num_leapfrog_steps=2,
          store_parameters_in_results=True),
        trace_fn=None)
    return states

_ = f(state)
```
The problem is with `nuts`:
```python
@tf.function(autograph=False)
def f(state_):
    states = tfp.mcmc.sample_chain(
        num_results=2,
        num_burnin_steps=2,
        current_state=state_,
        kernel=tfp.mcmc.NoUTurnSampler(
          target_log_prob_fn=logpfn,
          step_size=0.5),
        trace_fn=None)
    return states

_ = f(state)
```
What could be the real cause of the problem, I diggged into the code a little but wasn't fortunate enough to find anything, so a little hint would be great.

I'm not putting the traceback here because of the uncaught internal exception.
```
tf: 2.0.0
tfp: 0.8.0
```

**Comment by brianwa84** [comment]
Drive by comment:
If you know you have the same type of distribution in your mixture
components I'd recommend MixtureSameFamily.

**Comment by junpenglao** [comment]
I cannot reproduce it in TF/TFP nightly: https://colab.research.google.com/drive/1dhgKROL7UNFHu79vLSSZDKJlIUP95oXd

**Comment by rrkarim** [comment]
@jaingaurav you are importing tf-nightly here, I'm not sure if I can import last tfp-nightly with the stable tf version.

**Comment by junpenglao** [comment]
Hmm what error you are seeing in that case?

**Comment by rrkarim** [comment]
This one:
`
ValueError: Input tensor 'Placeholder_1:0' enters the loop with shape (1,), but has shape (None,) after one iteration. To allow the shape to vary across iterations, use the `shape_invariants` argument of tf.while_loop to specify a less-specific shape.
`

**Comment by junpenglao** [comment]
You mean using `state = tf.zeros(1)` right? It seems this breaks both `hmc` and `nuts`. 

Seems to me this is a bug using `tf.vectorized_map`, in this example you can do:
```python
likelihood = bimix_gauss
def vectorize(logpfn):
    def vectorize_logpfn(*state):
        return tf.vectorized_map(lambda mini_state: logpfn(*mini_state), state)
    return vectorize_logpfn

logpfn = likelihood.log_prob
logpfn_ = vectorize(logpfn)

# state = tf.zeros(5)
state = tf.zeros(1)
logpfn_(state), logpfn(state)
```
while both `logpfn_` and `logpfn` returns the same output, `logpfn_` breaks the sampler.

**Comment by rrkarim** [comment]
Is it a desired logic or it should be changed/fixed? I fill that it would be much more easier for a user to not worry about the batch size.

**Comment by junpenglao** [comment]
Definitively. cc @davmre as he is more experience with vectorized_map.

**Comment by rrkarim** [comment]
Just opened issue: tensorflow/tensorflow#35073, could be related.

**Comment by rrkarim** [comment]
Should be fixed by tensorflow/tensorflow#35162.
